### PR TITLE
feat: Caching and other efficiency improvements for mcap polyfill

### PR DESCRIPTION
**Pull request by dave-nicholas**
This PR caches the result of `requestMediaKeySystemAccess` saving time on subsequent calls.
It also makes the calls to `decodingInfo` synchronous. The reason for this, is the result of testing on multiple devices that the behaviour of `requestMediaKeySystemAccess` appears to be synchronous, making this synchronous can save over a second on older TVs.

The issue is currently assigned to my colleague Hugo Lima.
Closes https://github.com/shaka-project/shaka-player/issues/4574

### CODE REVIEW

File: lib/util/stream_utils.js [hunk 1]
```diff
@@ -532,10 +531,10 @@ shaka.util.StreamUtils = class {
           variant, usePersistentLicenses, srcEquals);
 
       for (const config of decodingConfigs) {
-        operations.push(getVariantDecodingInfos(variant, config));
+        // eslint-disable-next-line no-await-in-loop
+        await getVariantDecodingInfos(variant, config);
```
Comment (joeyparrish) [inline lib/util/stream_utils.js:535]: The reason eslint has a check for this is because serializing these async calls is bound to be slower than doing them in parallel.  Why did you change this?

File: lib/polyfill/media_capabilities.js [hunk 2]
```diff
@@ -24,6 +24,7 @@ shaka.polyfill.MediaCapabilities = class {
    * @export
    */
   static install() {
+    shaka.polyfill.MediaCapabilities.memoizedMediaKeySystemAccessRequests_ = {};
```
Comment (joeyparrish) [inline lib/polyfill/media_capabilities.js:27]: Please move this declaration outside the install() function, and give it a type annotation.

File: lib/polyfill/media_capabilities.js [hunk 3]
```diff
@@ -201,6 +217,34 @@ shaka.polyfill.MediaCapabilities = class {
...
   }
+
+  /**
+   * @param {!string} videoCodec
+   * @param {!string} audioCodec
+   * @param {!string} keySystem
+   * @return {!string}
+   * @static
+   */
+  static generateKeySystemCacheKey(videoCodec, audioCodec, keySystem) {
```
Comment (joeyparrish) [inline lib/polyfill/media_capabilities.js:228]: Please annotate this with `@private` and end the name with an underscore.

File: test/polyfill/media_capabilities_unit.js [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+      expect(isTypeSupportedSpy).toHaveBeenCalledTimes(2);
+      expect(isTypeSupportedSpy).toHaveBeenCalledWith(
+          mockDecodingConfig.video.contentType,
+      );
+      expect(isTypeSupportedSpy).toHaveBeenCalledWith(
+          mockDecodingConfig.audio.contentType,
+      );
+    });
+    it('should check codec support when MediaDecodingConfiguration ' +
```
Comment (joeyparrish) [inline test/polyfill/media_capabilities_unit.js:94]: style: Please leave a blank line between test cases

File: test/polyfill/media_capabilities_unit.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    it('should define decoding info method', () => {
+      shaka.polyfill.MediaCapabilities.install();
+
+      expect(navigator.mediaCapabilities.decodingInfo).toBeDefined();
+    });
+  });
+
+  describe('decodingInfo', () => {
+    it('should check codec support when MediaDecodingConfiguration ' +
+      'when type is "media-source"', () => {
```
Comment (joeyparrish) [inline test/polyfill/media_capabilities_unit.js:80]: What do you mean by "should check ... when MediaDecodingConfiguration when type is ..."?  Is there a typo or something?

File: lib/polyfill/media_capabilities.js [hunk 6]
```diff
@@ -201,6 +217,34 @@ shaka.polyfill.MediaCapabilities = class {
...
+  }
+
+  /**
+   * @param {!string} videoCodec
+   * @param {!string} audioCodec
+   * @param {!string} keySystem
+   * @param {!MediaKeySystemAccess} result
+   * @static
+   */
+  static appendKeySystemCache(videoCodec, audioCodec, keySystem, result) {
```
Comment (joeyparrish) [inline lib/polyfill/media_capabilities.js:239]: This is only used in a test that exercises it, but not elsewhere in the polyfill.

It appears the test in `describe('appendKeySystemCache'` is really a test of caching in `decodingInfo`.  Please move that test case to `describe('decodingInfo'` and migrate this function to a helper in the test itself.  You can use `@suppress {accessControls}` in the test helper annotations to access private members for the purposes of a test.

**Review by joeyparrish — CHANGES_REQUESTED** [review]

File: lib/util/stream_utils.js (hunk 1 above)
Comment (dave-nicholas) [reply lib/util/stream_utils.js:535]: The reason for this change is the result of testing across a couple of devices it yielded better performance. Across browsers, TVs and Chromecast. I saw no real time improvement between between `Promise.all` and `await` in a loop as `requestMediaKeySystemAccess` appears to not benefit from a multithreaded approach. 

I performed these tests with one of our assets that has many variants.
```
Browser
(promise.all)
56ms-75ms async
73ms-84ms cache + async
60ms-74ms cache + pre-cache async

(awaiting in a loop)
58ms -71ms sync 
40ms-54ms cache sync
35ms-46ms cache + pre-cache sync

Vizio TV
(promise.all)
4665 - 5940 async

Async + cache
4288 - 4576 (**This is the time on a secondary load as the initial 
doesn't benefit from the cache as all threads appear to stack and 
wait for `requestMediaKeySystemAccess`  calls to resolve synchronously**)

Sync
4594 - 5753

Sync + cache
3445 - 3615 (** gains benefit from the caching on initial play throughs **) 
```

I am happy to revert this part if you see different results, as I appreciate you will be aware of more use cases and devices.

File: lib/polyfill/media_capabilities.js (hunk 2 above)
Comment (dave-nicholas) [reply lib/polyfill/media_capabilities.js]: :+1:

File: lib/polyfill/media_capabilities.js (hunk 3 above)
Comment (dave-nicholas) [reply lib/polyfill/media_capabilities.js]: :+1:

File: test/polyfill/media_capabilities_unit.js (hunk 5 above)
Comment (dave-nicholas) [reply test/polyfill/media_capabilities_unit.js]: This is a typo sorry.

File: lib/polyfill/media_capabilities.js (hunk 6 above)
Comment (dave-nicholas) [reply lib/polyfill/media_capabilities.js:238]: The intention with `appendKeySystemCache` was to provide an api with which to pre-populate the cache externally.

Thanks again @joeyparrish for reviewing my PR :heart:.

File: lib/util/stream_utils.js (hunk 1 above)
Comment (sky-hugolima) [reply lib/util/stream_utils.js:535]: Are we sure there's a benefit in running requestMediaKeySystemAccess in parallel ?
I would agree to parallelise it, if it was network bound, but initialising the CDM is probably CPU bound instead and potentially even protected from running in multiple threads... which would explain the results we see above.

**Comment by avelad** [comment]
@dave-nicholas there are errors in the tests

**Comment by dave-nicholas** [comment]
> @dave-nicholas there are errors in the tests

Fixed 👍

File: lib/util/stream_utils.js (hunk 1 above)
Comment (joeyparrish) [reply lib/util/stream_utils.js:535]: Ah, okay.  It's possible that there is some startup state that has to be computed, and serializing queries may allow that state to be reused in subsequent queries.  Imagine this:

```
  Query A        Query B        Query C
     |              |              |
  Slow           Slow           Slow
  platform       platform       platform
  check          check          check
(cache miss)   (cache miss)   (cache miss)
```

vs this:

```
  Query A
     |
  Slow
  platform
  check
(cache miss)

  Query B
     |
  Cached
  platform
  check

  Query C
     |
  Cached
  platform
  check
```

Purely hypothetical.  I have no idea how those queries are handled by the device.

**Review by joeyparrish — CHANGES_REQUESTED** [review]

File: lib/polyfill/media_capabilities.js [hunk 7]
```diff
@@ -96,7 +96,7 @@ shaka.polyfill.MediaCapabilities = class {
       return res;
     }
 
-    if (mediaDecodingConfig.type == 'media-source') {
+    if (mediaDecodingConfig.type === 'media-source') {
```
Comment (avelad) [inline lib/polyfill/media_capabilities.js:99]: In all the code, we use == instead of ===, please revert this change.

File: lib/polyfill/media_capabilities.js [hunk 8]
```diff
@@ -116,7 +116,7 @@ shaka.polyfill.MediaCapabilities = class {
           return res;
         }
       }
-    } else if (mediaDecodingConfig.type == 'file') {
+    } else if (mediaDecodingConfig.type === 'file') {
```
Comment (avelad) [inline lib/polyfill/media_capabilities.js:119]: In all the code, we use == instead of ===, please revert this change.

**Review by avelad — CHANGES_REQUESTED** [review]

File: test/polyfill/media_capabilities_unit.js [hunk 9]
```diff
@@ -143,13 +155,14 @@ describe('MediaCapabilities', () => {
         });
   });
 
-  describe('generateKeySystemCacheKey', () => {
+  describe('generateKeySystemCacheKey_', () => {
```
[the remark this answers is not in the document]
Comment (dave-nicholas) [reply test/polyfill/media_capabilities_unit.js]: Removed 👍

File: lib/polyfill/media_capabilities.js (hunk 6 above)
Comment (dave-nicholas) [reply lib/polyfill/media_capabilities.js]: I have removed it.

File: lib/polyfill/media_capabilities.js (hunk 7 above)
Comment (dave-nicholas) [reply lib/polyfill/media_capabilities.js]: Done!

File: lib/util/stream_utils.js (hunk 1 above)
Comment (dave-nicholas) [reply lib/util/stream_utils.js:538]: I have added the requested comment.

**Review by avelad — APPROVED** [review]

**Review by joeyparrish — APPROVED** [review]
